### ISSUE TITLE
IDLE should restart instead of hanging when subprocess exits

### ISSUE THREAD

**Issue by terryjreedy** [python/cpython#82876]
BPO | [38695](https://bugs.python.org/issue38695)
--- | :---
Nosy | @terryjreedy
PRs | <li>python/cpython#17152</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/terryjreedy'
closed_at = None
created_at = <Date 2019-11-05.07:39:37.434>
labels = ['expert-IDLE', 'type-bug']
title = 'IDLE should restart instead of hanging when subprocess exits'
updated_at = <Date 2019-11-14.10:52:49.556>
user = 'https://github.com/terryjreedy'
```

bugs.python.org fields:
```python
activity = <Date 2019-11-14.10:52:49.556>
actor = 'taleinat'
assignee = 'terry.reedy'
closed = False
closed_date = None
closer = None
components = ['IDLE']
creation = <Date 2019-11-05.07:39:37.434>
creator = 'terry.reedy'
dependencies = []
files = []
hgrepos = []
issue_num = 38695
keywords = ['patch']
message_count = 1.0
messages = ['356009']
nosy_count = 1.0
nosy_names = ['terry.reedy']
pr_nums = ['17152']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue38695'
versions = []
```

</p></details>

**Comment by furkanonder** [comment]
@terryjreedy The issue seems to have been [resolved](https://github.com/python/cpython/pull/17152). I think we can close the issue.

**Comment by terryjreedy** [comment]
Not quite.  This issue is about not hanging for any other cause.

**Comment by serhiy-storchaka** [comment]
The case of the user process dying is already handled: an uncaught exception in its socket thread goes to `MyRPCServer.handle_error()`, which interrupts the main thread and exits, the socket is closed, and `poll_subprocess()` in the GUI process restarts the Shell. I verified this by killing the user process with SIGKILL. An outer try/except in run.py would not add anything here.

The remaining hang is when the user process stays alive but does not respond. Synchronous requests (calltips, completions, the debugger and stack viewer) run user code in the socket thread of the user process: `__call__`, `inspect.signature()`, `dir()`, `__getattr__()`, `repr()`. If that code blocks, the GUI process waits for the response forever in `rpc.SocketIO._getresponse()`, the Tk event loop is not running, so even Ctrl-C does not work. For example:

```python
class A:
    @property
    def __call__(self):
        import time; time.sleep(60)
a = A()
```

Typing `a(` in the Shell freezes IDLE for a minute. #17152 only fixed a particular exception which wedged the socket thread in the same way.

The fix would be a timeout for synchronous requests from the GUI process (`remotecall()` → `_getresponse()`, which already polls in 0.05 s steps). Since the socket thread of the user process is still stuck after the timeout (user output is blocked too), the GUI should then restart the user process as after a crash, with a message in the Shell. Open questions are the timeout value (a legitimately slow `__dir__()` or `__call__` would be misreported) and whether debugger requests should have a timeout at all.

**Comment by serhiy-storchaka** [comment]
It is better to not restart the subprocess immediately, but just fix non-responsibility of the GUI, indicate non-responsibility of the subprocess, and allow the user to wait and restart manually. IDLE induced remote calls (for calltips, completion) could be terminated immediately, but this is a complex task.

**Comment by terryjreedy** [comment]
'indicate non-responsibiliy': does this mean a popup message?

**Comment by serhiy-storchaka** [comment]
No, a status bar indicator. The prototype shows "⏳ Not responding" in the Shell status bar when a calltip or completion request times out, clears it as soon as anything is received from the user process, and abandons the timed out request without restarting. A popup would interrupt the user, and the subprocess usually catches up. If the user decides it is stuck, Shell → Restart Shell as now. I am still working on it.
